storage/scylla_api.c

```diff
@@ -0,0 +1,707 @@
+                                             size_t trans_num) {
+  status_t ret = SC_OK;
+  const CassPrepared* insert_prepared = NULL;
+  CassStatement* statement = NULL;
+  const char* insert_query =
+      "INSERT INTO permanent.bundleTable (bundle, address, hash, message, value, timestamp, trunk, branch)"
+      "VALUES (?, ?, ?, ?, ?, ?, ?, ?);";
+  if (session == NULL || transaction == NULL || trans_num == 0) {
+    return SC_STORAGE_INVAILD_INPUT;
+  }
+
+  if (prepare_query(session, insert_query, &insert_prepared) != CASS_OK) {
+    ta_log_error("%s\n", "prepare INSERT query fail");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+  for (size_t i = 0; i < trans_num; i++) {
+    statement = ret_insert_bundle_statement(insert_prepared, transaction + i);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t insert_transaction_into_edgeTable(CassSession* session, scylla_iota_transaction_t* transaction,
+                                           size_t trans_num) {
+  status_t ret = SC_OK;
+  const CassPrepared* insert_prepared = NULL;
+  CassStatement* statement = NULL;
+  const char* insert_query =
+      "INSERT INTO permanent.edgeTable (edge, bundle, address, hash)"
+      "VALUES (?, ?, ?, ?);";
+
+  if (transaction == NULL || trans_num == 0) {
+    return SC_STORAGE_INVAILD_INPUT;
+  }
+
+  if (prepare_query(session, insert_query, &insert_prepared) != CASS_OK) {
+    ta_log_error("%s\n", "prepare INSERT query fail");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  for (size_t i = 0; i < trans_num; i++) {
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->address);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->trunk);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->branch);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t select_from_bundleTable(CassSession* session, CassStatement* statement,
+                                 scylla_iota_transaction_t** transaction, size_t* transactionNum) {
+  int idx;
+  status_t ret = SC_OK;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
```

Comment:
logger ?

Reply:
print_error use ta_logger 
```
static void print_error(CassFuture* future) {
  const char* message;
  size_t message_length;
  cass_future_error_message(future, &message, &message_length);
  ta_log_error("Error: %.*s\n", (int)message_length, message);
}
```
And there's corresponing error message in the caller function.